bellman/src/domain/fft.rs

```diff
@@ -0,0 +1,177 @@
+use crate::{multicore, arith};
+use multicore::Worker;
+use arith::Group;
+
+use ff::{Field, ScalarEngine};
+
+pub fn run_optimal_fft<E,T>(a: &mut [T], worker: &Worker, omega: &E::Fr, log_n: u32) 
+where
+    E: ScalarEngine,
```

Comment:
Please look into more meaningful parameters if possible

Reply:
Agreed, I haven't changed the underlying fft until I can find the original source of their work. This looks like a pseudo-code implementation of a 'parallel' and 'serial' fft, hence the 'omega'.

In general, I want to overhaul the worker design pattern and because it seems like a cluttered way to implement multi-threaded computations, I might look into abstracting a macro over the top. I will rename 'a' into something like sequence, which is the standard nomenclature for fft.

Thoughts on something like this?

`pub fn run_optimal_fft<E,T>(sequence: &mut [T], omega: &E::Fr, log_n: u32) 
where
    E: ScalarEngine,
    for <'a> T: Group<'a,E> 
{
    let log_cpus = worker.log_num_cpus();

    if log_n <= log_cpus {
        serial_fft(sequence, omega, log_n);
    } else {
        parallel_fft(sequence, omega, log_n, log_cpus);
    }
}
`

Currently, multi-threading looks like this:
`worker.scope(0, |scope, _| {
        let a: &[T] = a;

        for (j, tmp) in tmp.iter_mut()
            .enumerate() 
        {
            scope.spawn(move || {
                /** Code Block **/
            });
        }
    });
`
There's a lot of clutter and I want it instead to look like this:

` multi_thread!(
      /** Code Block **/
)
`

This should remove the worker pattern, which is quite ambiguous until you see its implementation.

Reply:
The new design pattern looks like this:
```
    let tmp: _ = &tmp;
    multi_thread!(a.len(), chunk::init() => {
        for (idx, a) in a.chunks_mut(chunk).enumerate() => {
            let mut idx = idx * chunk;
            let mask = (1 << log_cpus) - 1;
            for a in a {
                *a = tmp[idx & mask][idx >> log_cpus];
                idx += 1;
            }
        }
    });
```

Reply:
I've added the `lazy_static` dependency exposed to the macro so that we have essentially removed the need for `Worker` on most structures. I think this should reduce the parameters in most functions and hopefully has reduced ambiguity for the `fft` module.

Reply:
Resolved, good work